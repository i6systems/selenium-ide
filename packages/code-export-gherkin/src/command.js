// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { codeExport as exporter } from '@seleniumhq/side-utils'

let activeKeyword

export const emitters = {
  addSelection: skip,
  and: emitAnd,
  answerOnNextPrompt: skip,
  assert: skip,
  assertAlert: skip,
  assertChecked: skip,
  assertConfirmation: skip,
  assertEditable: skip,
  assertElementPresent: skip,
  assertElementNotPresent: skip,
  assertNotChecked: skip,
  assertNotEditable: skip,
  assertNotSelectedValue: skip,
  assertNotText: skip,
  assertPrompt: skip,
  assertSelectedLabel: skip,
  assertSelectedValue: skip,
  assertValue: skip,
  assertText: skip,
  assertTitle: skip,
  check: skip,
  chooseCancelOnNextConfirmation: skip,
  chooseCancelOnNextPrompt: skip,
  chooseOkOnNextConfirmation: skip,
  click: skip,
  clickAt: skip,
  close: skip,
  debugger: skip,
  do: skip,
  doubleClick: skip,
  doubleClickAt: skip,
  dragAndDropToObject: skip,
  echo: skip,
  editContent: skip,
  else: skip,
  elseIf: skip,
  end: skip,
  executeScript: skip,
  executeAsyncScript: skip,
  forEach: skip,
  given: emitGiven,
  if: skip,
  mouseDown: skip,
  mouseDownAt: skip,
  mouseMove: skip,
  mouseMoveAt: skip,
  mouseOver: skip,
  mouseOut: skip,
  mouseUp: skip,
  mouseUpAt: skip,
  open: skip,
  pause: skip,
  removeSelection: skip,
  repeatIf: skip,
  run: skip,
  runScript: skip,
  select: skip,
  selectFrame: skip,
  selectWindow: skip,
  sendKeys: skip,
  setSpeed: skip,
  setWindowSize: skip,
  store: skip,
  storeAttribute: skip,
  storeJson: skip,
  storeText: skip,
  storeTitle: skip,
  storeValue: skip,
  storeWindowHandle: skip,
  storeXpathCount: skip,
  submit: skip,
  then: emitThen,
  times: skip,
  type: skip,
  uncheck: skip,
  verify: skip,
  verifyChecked: skip,
  verifyEditable: skip,
  verifyElementPresent: skip,
  verifyElementNotPresent: skip,
  verifyNotChecked: skip,
  verifyNotEditable: skip,
  verifyNotSelectedValue: skip,
  verifyNotText: skip,
  verifySelectedLabel: skip,
  verifySelectedValue: skip,
  verifyText: skip,
  verifyTextContains: skip,
  verifyTitle: skip,
  verifyValue: skip,
  waitForElementEditable: skip,
  waitForElementPresent: skip,
  waitForElementVisible: skip,
  waitForElementNotEditable: skip,
  waitForElementNotPresent: skip,
  waitForElementNotVisible: skip,
  waitForText: skip,
  waitForTextContains: skip,
  webdriverAnswerOnVisiblePrompt: skip,
  webdriverChooseCancelOnVisibleConfirmation: skip,
  webdriverChooseCancelOnVisiblePrompt: skip,
  webdriverChooseOkOnVisibleConfirmation: skip,
  when: emitWhen,
  while: skip,
}

exporter.register.preprocessors(emitters)

function init() {
  activeKeyword = null
}

function register(command, emitter) {
  exporter.register.emitter({ command, emitter, emitters })
}

function emit(command) {
  return exporter.emit.command(command, emitters[command.command], {
    variableLookup,
    emitNewWindowHandling,
  })
}

function canEmit(commandName) {
  return !!emitters[commandName]
}

function variableLookup() {
  return ''
}

function emitWaitForWindow() {
  const generateMethodDeclaration = _ => {
    return ''
  }
  const commands = [{ level: 0, statement: '' }]
  return Promise.resolve({
    name: 'waitForWindow',
    commands,
    generateMethodDeclaration,
  })
}

async function emitNewWindowHandling() {
  return Promise.resolve(undefined)
}

async function emitAnd(step) {
  if (activeKeyword === null) {
    return Promise.reject(
      new Error('"And" cannot be the first step in an scenario')
    )
  }
  return emitStep('And', step)
}

async function emitGiven(step) {
  if (activeKeyword !== null) {
    return Promise.reject(
      new Error('"Given" must be the first step in an scenario')
    )
  }
  if (activeKeyword === 'Given') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "Given" steps, use an "And" step instead'
      )
    )
  }
  return emitStep('Given', step)
}

async function emitThen(step) {
  if (activeKeyword === 'Then') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "Then" steps, use an "And" step instead'
      )
    )
  }
  return emitStep('Then', step)
}

async function emitWhen(step) {
  if (activeKeyword === 'When') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "When" steps, use an "And" step instead'
      )
    )
  }
  return emitStep('When', step)
}

function emitStep(command, step) {
  let result = command + ' ' + step
  if (command !== 'And') {
    activeKeyword = command
  }
  return Promise.resolve(result)
}

function skip() {
  return Promise.resolve('')
}

export default {
  canEmit,
  emit,
  register,
  init,
  extras: { emitWaitForWindow, emitNewWindowHandling },
}
