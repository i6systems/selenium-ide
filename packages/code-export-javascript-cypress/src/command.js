// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { codeExport as exporter } from '@seleniumhq/side-utils'
import location from './location'

let activeKeyword

let emittedSteps = []

let emitting

export const emitters = {
  addSelection: emitSelect,
  and: emitAnd,
  answerOnNextPrompt: skip,
  assert: emitAssert,
  assertAlert: skip,
  assertChecked: emitVerifyChecked,
  assertConfirmation: skip,
  assertEditable: emitVerifyEditable,
  assertElementPresent: emitVerifyElementPresent,
  assertElementNotPresent: emitVerifyElementNotPresent,
  assertNotChecked: emitVerifyNotChecked,
  assertNotEditable: emitVerifyNotEditable,
  assertNotSelectedValue: emitVerifyNotSelectedValue,
  assertNotText: emitVerifyNotText,
  assertPrompt: skip,
  assertSelectedLabel: emitVerifySelectedLabel,
  assertSelectedValue: emitVerifyValue,
  assertValue: emitVerifyValue,
  assertText: emitVerifyText,
  assertTitle: emitVerifyTitle,
  check: emitCheck,
  chooseCancelOnNextConfirmation: skip,
  chooseCancelOnNextPrompt: skip,
  chooseOkOnNextConfirmation: skip,
  click: emitClick,
  clickAt: emitClick,
  close: skip,
  debugger: skip,
  do: skip,
  doubleClick: emitDoubleClick,
  doubleClickAt: emitDoubleClick,
  dragAndDropToObject: emitDragAndDrop,
  echo: emitEcho,
  editContent: emitEditContent,
  else: skip,
  elseIf: skip,
  end: skip,
  executeScript: skip,
  executeAsyncScript: skip,
  forEach: skip,
  given: emitGiven,
  if: skip,
  mouseDown: emitMouseDown,
  mouseDownAt: emitMouseDown,
  mouseMove: emitMouseMove,
  mouseMoveAt: emitMouseMove,
  mouseOver: emitMouseMove,
  mouseOut: emitMouseOut,
  mouseUp: emitMouseUp,
  mouseUpAt: emitMouseUp,
  open: emitOpen,
  pause: emitPause,
  removeSelection: emitSelect,
  repeatIf: skip,
  run: emitRun,
  runScript: skip,
  select: emitSelect,
  selectFrame: skip,
  selectWindow: skip,
  sendKeys: emitSendKeys,
  setSpeed: skip,
  setWindowSize: emitSetWindowSize,
  store: emitStore,
  storeAttribute: emitStoreAttribute,
  storeJson: emitStoreJson,
  storeText: emitStoreText,
  storeTitle: emitStoreTitle,
  storeValue: emitStoreValue,
  storeWindowHandle: skip,
  storeXpathCount: emitStoreXpathCount,
  submit: emitSubmit,
  then: emitThen,
  times: skip,
  type: emitSendKeys,
  uncheck: emitUncheck,
  verify: emitAssert,
  verifyChecked: emitVerifyChecked,
  verifyEditable: emitVerifyEditable,
  verifyElementPresent: emitVerifyElementPresent,
  verifyElementNotPresent: emitVerifyElementNotPresent,
  verifyNotChecked: emitVerifyNotChecked,
  verifyNotEditable: emitVerifyNotEditable,
  verifyNotSelectedValue: emitVerifyNotSelectedValue,
  verifyNotText: emitVerifyNotText,
  verifySelectedLabel: emitVerifySelectedLabel,
  verifySelectedValue: emitVerifyValue,
  verifyText: emitVerifyText,
  verifyTitle: emitVerifyTitle,
  verifyValue: emitVerifyValue,
  waitForElementEditable: emitVerifyEditable,
  waitForElementPresent: emitVerifyElementPresent,
  waitForElementVisible: emitWaitForElementVisible,
  waitForElementNotEditable: emitVerifyNotEditable,
  waitForElementNotPresent: emitVerifyElementNotPresent,
  waitForElementNotVisible: emitWaitForElementNotVisible,
  webdriverAnswerOnVisiblePrompt: skip,
  waitForText: emitVerifyText,
  webdriverChooseCancelOnVisibleConfirmation: skip,
  webdriverChooseCancelOnVisiblePrompt: skip,
  webdriverChooseOkOnVisibleConfirmation: skip,
  when: emitWhen,
  while: skip,
}

exporter.register.preprocessors(emitters)

function init() {
  activeKeyword = null
  emitting = true
}

function register(command, emitter) {
  exporter.register.emitter({ command, emitter, emitters })
}

function emit(command) {
  if (
    command.command !== 'and' &&
    command.command !== 'given' &&
    command.command !== 'then' &&
    command.command !== 'when' &&
    emitting === false
  ) {
    return skip()
  }

  return exporter.emit.command(command, emitters[command.command], {
    variableLookup,
    emitNewWindowHandling,
  })
}

function canEmit(commandName) {
  return !!emitters[commandName]
}

function variableLookup(varName) {
  return `vars["${varName}"]`
}

function variableSetter(varName, value) {
  return varName ? `vars["${varName}"] = ${value}` : ''
}

function emitWaitForWindow() {
  return skip()
}

async function emitNewWindowHandling() {
  return skip()
}

function emitAssert(varName, value) {
  return Promise.resolve(`expect(vars['${varName}']).to.equal('${value}')`)
}

async function emitCheck(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.check({ force:true })`
  )
}

async function emitClick(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.click({ force:true })`
  )
}

async function emitDoubleClick(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.dblclick({ force:true })`
  )
}

async function emitDragAndDrop(dragged, dropped) {
  const commands = [
    { level: 0, statement: `{` },
    {
      level: 1,
      statement: `${await location.emit(dragged)}.trigger('dragstart')`,
    },
    {
      level: 1,
      statement: `${await location.emit(dropped)}.trigger('drop')`,
    },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitEcho(message) {
  const _message = message.startsWith('vars[') ? message : `"${message}"`
  return Promise.resolve(`cy.log(${_message})`)
}

async function emitEditContent(locator, content) {
  return Promise.resolve(`
    ${await location.emit(
      locator
    )}.then($element => { $element.innerText = '${content}' }"`)
}

async function emitMouseDown(locator) {
  return Promise.resolve(`${await location.emit(locator)}.trigger('mousedown')`)
}

async function emitMouseMove(locator) {
  return Promise.resolve(`${await location.emit(locator)}.trigger('mousemove')`)
}

async function emitMouseOut(locator) {
  return Promise.resolve(`
    ${await location.emit(locator)}.trigger('mouseleave')`)
}

async function emitMouseUp(locator) {
  return Promise.resolve(`${await location.emit(locator)}.trigger('mouseup')`)
}

function emitOpen(target) {
  const url = /^(file|http|https):\/\//.test(target)
    ? `"${target}"`
    : `"${global.baseUrl}${target}"`
  return Promise.resolve(`cy.visit(${url})`)
}

async function emitPause(time) {
  const commands = [{ level: 0, statement: `cy.wait(${time})` }]
  return Promise.resolve({ commands })
}

async function emitRun(testName) {
  return Promise.resolve(`${exporter.parsers.sanitizeName(testName)}()`)
}

async function emitSetWindowSize(size) {
  const [width, height] = size.split('x')
  return Promise.resolve(`cy.viewport(${width}, ${height})`)
}

async function emitSelect(selectElement, option) {
  if (option.startsWith('label=')) {
    option = option.substring(6)
  }
  return Promise.resolve(
    `${await location.emit(selectElement)}.select('${option}', { force:true })`
  )
}

function generateSendKeysInput(value) {
  if (typeof value === 'object') {
    return value
      .map(s => {
        if (s.startsWith('vars[')) {
          return s
        } else if (s.startsWith('Key[')) {
          const key = s.match(/\['(.*)'\]/)[1]
          return `Key.${key}`
        } else {
          return `"${s}"`
        }
      })
      .join(', ')
  } else {
    if (value.startsWith('vars[')) {
      return value
    } else {
      return `"${value}"`
    }
  }
}

async function emitSendKeys(locator, value) {
  return Promise.resolve(
    `${await location.emit(locator)}.type(${generateSendKeysInput(
      value
    )}, { force:true })`
  )
}

async function emitStore(value, varName) {
  return Promise.resolve(variableSetter(varName, `"${value}"`))
}

async function emitStoreAttribute(locator, varName) {
  const attributePos = locator.lastIndexOf('@')
  const elementLocator = locator.slice(0, attributePos)
  const attributeName = locator.slice(attributePos + 1)

  const commands = [
    {
      level: 0,
      statement: `${await location.emit(
        elementLocator
      )}.its('${attributeName}').then(value => {`,
    },
    { level: 1, statement: `${variableSetter(varName, 'value')}` },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitStoreJson(json, varName) {
  return Promise.resolve(variableSetter(varName, `JSON.parse('${json}')`))
}

async function emitStoreText(locator, varName) {
  const commands = [
    {
      level: 0,
      statement: `${await location.emit(locator)}.text().then(text => {`,
    },
    { level: 1, statement: `${variableSetter(varName, 'text')}` },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitStoreTitle(_, varName) {
  const commands = [
    {
      level: 0,
      statement: `cy.title().then(title => {`,
    },
    { level: 1, statement: `${variableSetter(varName, 'title')}` },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitStoreValue(locator, varName) {
  const commands = [
    {
      level: 0,
      statement: `${await location.emit(locator)}.its('value').then(value => {`,
    },
    { level: 1, statement: `${variableSetter(varName, 'value')}` },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitStoreXpathCount(locator, varName) {
  const commands = [
    {
      level: 0,
      statement: `${await location.emit(
        locator
      )}.its('length').then(count => {`,
    },
    { level: 1, statement: `${variableSetter(varName, 'count')}` },
    { level: 0, statement: `}` },
  ]
  return Promise.resolve({ commands })
}

async function emitSubmit(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.submit({ force:true })`
  )
}

async function emitUncheck(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.uncheck({ force:true })`
  )
}

async function emitVerifyChecked(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.should('have.prop', 'checked')`
  )
}

async function emitVerifyEditable(locator) {
  return Promise.resolve(`${await location.emit(locator)}.should('be.enabled')`)
}

async function emitVerifyElementPresent(locator) {
  return Promise.resolve(`${await location.emit(locator)}.should('exist')`)
}

async function emitVerifyElementNotPresent(locator) {
  return Promise.resolve(`${await location.emit(locator)}.should('not.exist')`)
}

async function emitVerifyNotChecked(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.should('not.have.prop', 'checked')`
  )
}

async function emitVerifyNotEditable(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.should('not.be.enabled')`
  )
}

async function emitVerifyNotSelectedValue(locator, expectedValue) {
  return Promise.resolve(
    `${await location.emit(
      locator
    )}.should('not.have.value','${expectedValue}')`
  )
}

async function emitVerifyNotText(locator, text) {
  return Promise.resolve(
    `${await location.emit(
      locator
    )}.text().should('not.equal','${exporter.emit.text(text)}')`
  )
}

async function emitVerifySelectedLabel(locator, labelValue) {
  return Promise.resolve(
    `${await location.emit(
      locator
    )}.find(':selected').should('contain','${labelValue}')`
  )
}

async function emitVerifyText(locator, text) {
  return Promise.resolve(
    `${await location.emit(
      locator
    )}.text().should('equal','${exporter.emit.text(text)}')`
  )
}

async function emitVerifyValue(locator, value) {
  return Promise.resolve(
    `${await location.emit(locator)}.should('have.value','${value}')`
  )
}

async function emitVerifyTitle(title) {
  return Promise.resolve(`cy.title().should('equal','${title}')`)
}

function skip() {
  return Promise.resolve(undefined)
}

async function emitWaitForElementVisible(locator) {
  return Promise.resolve(`${await location.emit(locator)}.should('be.visible')`)
}

async function emitWaitForElementNotVisible(locator) {
  return Promise.resolve(
    `${await location.emit(locator)}.should('not.be.visible')`
  )
}

async function emitAnd(step) {
  if (activeKeyword === null) {
    return Promise.reject(
      new Error('"And" cannot be the first step in an scenario')
    )
  }
  return emitGherkinStep(activeKeyword, step)
}

async function emitGiven(step) {
  if (activeKeyword !== null) {
    return Promise.reject(
      new Error('"Given" must be the first step in an scenario')
    )
  }
  if (activeKeyword === 'Given') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "Given" steps, use an "And" step instead'
      )
    )
  }
  return emitGherkinStep('Given', step)
}

async function emitThen(step) {
  if (activeKeyword === 'Then') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "Then" steps, use an "And" step instead'
      )
    )
  }
  return emitGherkinStep('Then', step)
}

async function emitWhen(step) {
  if (activeKeyword === 'When') {
    return Promise.reject(
      new Error(
        'You cannot have two consecutive "When" steps, use an "And" step instead'
      )
    )
  }
  return emitGherkinStep('When', step)
}

async function emitGherkinStep(keyword, step) {
  if (step in emittedSteps) {
    emitting = false
    return skip()
  }
  emitting = true
  emittedSteps[step] = true

  let commands = []
  if (activeKeyword !== null) {
    commands.push({ level: 0, statement: `})` })
  }
  activeKeyword = keyword
  commands.push({ level: 0, statement: `` })
  commands.push({ level: 0, statement: `${keyword}(\`${step}\`, () => {` })

  return Promise.resolve({
    commands: commands,
    setStartingLevel: 0,
    endingLevelAdjustment: 1,
  })
}

export default {
  canEmit,
  emit,
  register,
  init,
  extras: { emitWaitForWindow },
}
