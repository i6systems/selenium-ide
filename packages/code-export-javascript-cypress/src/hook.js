// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the specific language governing permissions and limitations
// under the License.

import { codeExport as exporter } from '@seleniumhq/side-utils'

const emitters = {
  afterAll: empty,
  afterEach: empty,
  beforeAll: empty,
  beforeEach: empty,
  declareDependencies,
  declareMethods: empty,
  declareVariables: empty,
  inEachBegin: empty,
  inEachEnd: empty,
}

function generate(hookName) {
  return new exporter.hook(emitters[hookName]())
}

export function generateHooks() {
  let result = {}
  Object.keys(emitters).forEach(hookName => {
    result[hookName] = generate(hookName)
  })
  return result
}

function declareDependencies() {
  return {
    startingSyntax: {
      commands: [
        {
          level: 0,
          statement:
            'import { Before, Given, When, Then } from "cypress-cucumber-preprocessor/steps"',
        },
        {
          level: 0,
          statement: "const Database = require('../../support/database')",
        },
        {
          level: 0,
          statement: "const Redis = require('../../support/redis')",
        },
        {
          level: 0,
          statement: 'const Helpers = require("../../support/helpers")',
        },
        {
          level: 0,
          statement: '',
        },
        {
          level: 0,
          statement: 'Before(() => {',
        },
        {
          level: 1,
          statement: 'Database.reload()',
        },
        {
          level: 1,
          statement: 'Redis.clear()',
        },
        {
          level: 1,
          statement: "Helpers.login('test.user@example.com')",
        },
        {
          level: 0,
          statement: '})',
        },
      ],
    },
  }
}

function empty() {
  return {}
}
