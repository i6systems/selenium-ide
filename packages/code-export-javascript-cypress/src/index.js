// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { codeExport as exporter } from '@seleniumhq/side-utils'
import emitter from './command'
import location from './location'
import { generateHooks } from './hook'

// Define language options
export const displayName = 'JavaScript Cypress'

export let opts = {}
opts.emitter = emitter
opts.hooks = generateHooks()
opts.fileExtension = '.js'
opts.commandPrefixPadding = '    '
opts.terminatingKeyword = '})'
opts.commentPrefix = '//'
opts.testLevel = 0
opts.commandLevel = 1
opts.generateMethodDeclaration = generateMethodDeclaration

// Create generators for dynamic string creation of primary entities (e.g., filename, methods, test, and suite)
function generateTestDeclaration() {
  return ''
}

function generateMethodDeclaration() {
  return {
    body: '',
    terminatingKeyword: '',
  }
}

function generateSuiteDeclaration() {
  return ''
}

function generateFilename(name) {
  return `${exporter.parsers.uncapitalize(
    exporter.parsers.sanitizeName(name)
  )}${opts.fileExtension}`
}

// Emit an individual test, wrapped in a suite (using the test name as the suite name)
export async function emitTest({
  baseUrl,
  test,
  tests,
  project,
  enableOriginTracing,
  beforeEachOptions,
  enableDescriptionAsComment,
}) {
  global.baseUrl = baseUrl
  const testDeclaration = generateTestDeclaration(test.name)
  const result = await exporter.emit.test(test, tests, {
    ...opts,
    testDeclaration,
    enableOriginTracing,
    enableDescriptionAsComment,
    project,
  })
  const suiteName = test.name
  const suiteDeclaration = generateSuiteDeclaration(suiteName)
  const _suite = await exporter.emit.suite(result, tests, {
    ...opts,
    suiteDeclaration,
    suiteName,
    project,
    beforeEachOptions,
  })
  return {
    filename: generateFilename(test.name),
    body: emitOrderedSuite(_suite),
  }
}

// Emit a suite with all of its tests
export async function emitSuite({
  baseUrl,
  suite,
  tests,
  project,
  enableOriginTracing,
  beforeEachOptions,
  enableDescriptionAsComment,
}) {
  global.baseUrl = baseUrl
  const result = await exporter.emit.testsFromSuite(tests, suite, opts, {
    enableOriginTracing,
    generateTestDeclaration,
    enableDescriptionAsComment,
    project,
  })
  const suiteDeclaration = generateSuiteDeclaration(suite.name)
  const _suite = await exporter.emit.suite(result, tests, {
    ...opts,
    suiteDeclaration,
    suite,
    project,
    beforeEachOptions,
  })
  return {
    filename: generateFilename(suite.name),
    body: emitOrderedSuite(_suite),
  }
}

function emitOrderedSuite(emittedSuite) {
  let result = ''
  result += emittedSuite.dependencies
  result += emittedSuite.suiteDeclaration
  result += emittedSuite.variables
  result += emittedSuite.beforeAll
  result += emittedSuite.beforeEach
  result += emittedSuite.afterEach
  result += emittedSuite.afterAll
  result += emittedSuite.methods
  if (emittedSuite.tests.testDeclaration) {
    const test = emittedSuite.tests
    result += test.testDeclaration
    result += test.inEachBegin
    result += test.commands
    result += test.inEachEnd
    result += test.testEnd
  } else {
    for (const testName in emittedSuite.tests) {
      const test = emittedSuite.tests[testName]
      result += test.testDeclaration
      result += test.inEachBegin
      result += test.commands
      result += test.inEachEnd
      result += test.testEnd
    }
  }
  return result
}

export default {
  emit: {
    test: emitTest,
    suite: emitSuite,
    locator: location.emit,
  },
  register: {
    command: emitter.register,
    variable: opts.hooks.declareVariables.register,
    dependency: opts.hooks.declareDependencies.register,
    beforeAll: opts.hooks.beforeAll.register,
    beforeEach: opts.hooks.beforeEach.register,
    afterEach: opts.hooks.afterEach.register,
    afterAll: opts.hooks.afterAll.register,
    inEachBegin: opts.hooks.inEachBegin.register,
    inEachEnd: opts.hooks.inEachEnd.register,
  },
}
