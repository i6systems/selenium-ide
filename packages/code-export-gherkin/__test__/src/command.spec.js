// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Command from '../../src/command'
import {
  Commands,
  ControlFlowCommandNames,
} from '../../../selenium-ide/src/neo/models/Command'
import { opts } from '../../src/index'
import { codeExport as exporter } from '@seleniumhq/side-utils'

const commandPrefixPadding = opts.commandPrefixPadding

async function prettify(command, { fullPayload } = {}) {
  const commandBlock = await Command.emit(command)
  const result = exporter.prettify(commandBlock, {
    commandPrefixPadding,
  })
  return fullPayload ? result : result.body
}

describe('command code emitter', () => {
  it('should emit all known commands', () => {
    let result = []
    Commands.array.forEach(command => {
      if (!Command.canEmit(command)) {
        result.push(command)
      }
    })
    expect(() => {
      if (result.length) {
        if (result.length === 1) {
          throw new Error(`${result[0]} has no emitter, write one!`)
        } else {
          throw new Error(`No emitters for ${result.join(', ')}. Write them!`)
        }
      }
    }).not.toThrow()
  })
  it('should emit a command without comment', () => {
    const command = {
      command: 'addSelection',
      target: 'css=select',
      value: 'label=A label',
    }
    return expect(prettify(command)).resolves.toMatchSnapshot()
  })
  it('should emit a command with a comment', () => {
    const command = {
      command: 'assert',
      target: 'varrrName',
      value: 'blah',
      comment: 'test step',
    }
    return expect(prettify(command)).resolves.toMatchSnapshot()
  })
  it('should emit new window handling, if command opens a new window', () => {
    const command = {
      command: 'click',
      target: 'css=button',
      value: '',
      opensWindow: true,
      windowHandleName: 'newWin',
      windowTimeout: 2000,
    }
    return expect(prettify(command)).resolves.toMatchSnapshot()
  })
})
