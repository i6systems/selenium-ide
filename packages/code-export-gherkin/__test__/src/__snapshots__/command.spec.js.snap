// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command code emitter should emit a command with a comment 1`] = `"        test step"`;

exports[`command code emitter should emit a command without comment 1`] = `undefined`;

exports[`command code emitter should emit new window handling, if command opens a new window 1`] = `undefined`;
